### DISCUSSION TITLE
Installation Instructions

### DISCUSSION THREAD

**Discussion by Siff3D** [QuickSander/homebridge-mieleathome#86] [Show and tell]
There are only very basic installation instructions for this plugin and I want to share the steps I took to install and configure it. Hope that this will help somebody:

1. Download the "Miele app - Smart Home" app (if not done already)
2. Create a _Miele@Home_ account in the app (if not done already)
3. Go to the [https://www.miele.com/f/com/en/register_api.aspx](https://www.miele.com/f/com/en/register_api.aspx) page and create a _Miele Developer_ account:
- The app name doesn't matter - I used a made-up name. It seems that there could be only one app per account - when I registered using a new app name, the previous one disappeared from the "Miele app - Smart Home" app (see the note at the end) and the new app name showed up after I completed the rest of this step
- Use the same email as the one for the _Miele@Home_ account. I guess that this adds a "developer" role to the _Miele@Home_ account, but this is just a guess
- Click the "Register" button
- Wait for an email from "donotreply@mieleathome.com" to arrive (takes some time. I waited ~10 min.)
- Click on the activation link provided in the email
- Copy the **Client ID** and the **Client Secret** from the registration page. **Note** that they cannot be retrieved once the page is closed, so I would recommend taking a screenshot of the page before closing it just in case
- Check the "I agree to terms of use." check-box
- Click on the "Proceed" button
- The registration of the _Miele Developer_ account is now complete
4.  Install the plugin
5.  In the plugin configuration page enter the **Client ID** and the **Client Secret** from above
6.  Restart HomeBridge
7.  Open the plugin configuration
8.  Check that the "Authorization grant server online and reachable..." message is displayed in green
9.  Press "Next"
10.  In the page that opens enter the _Miele@Home_ credentials (the app credentials)
11. Accept the app authorization request (the app name will the the one entered during the _Miele Developer_ account registration)
12. Check that the "Miele@Home Homebridge plugin: Success" and the "Authorization grant successfully received" messages are displayed in green
13. Press the "Dismiss" button to close the dialog
14. Go back to the plugin settings page and press "Save"
15. Restart HomeBridge

Done.

**Note**: The app specified and given permission to access with your appliances as part of the registration of the _Miele Developer_ account will appear under "Third party apps" in the "Profile settings" in the Miele@Home app and its access can be revoked from there, if needed.

**Note**: Thanks to @keepers1221 for his post [here](https://github.com/QuickSander/homebridge-mieleathome/issues/82#issuecomment-1774009479), which I used as a starting point for these instructions.

**Comment by mark-fitzgerald** [comment]
I followed these instructions, but at step 14, I see two toast messages. One that says "Authorization grant received" and another one under it that says "Failed to request token. Request failed with status code 401". 
<img width="776" height="423" alt="image" src="https://github.com/user-attachments/assets/3508b747-ba05-4a7f-878f-e66db27f46fe" />

No additional information in the browser console log.
```
[Log] New token status received: "authorization_grant_received" (index.html, line 65)
[Log] Token retrieval error: {"message":"Request failed with status code 401","title":"Failed to request token"} (index.html, line 91)
```
How do I resolve this request failure?

**Comment by Siff3D** [comment]
Hi @mark-fitzgerald,

I don't recall seeing this before. Since I'm not the author of the plugin, I can only guess what might be the issue:

1.  An issue on Miele's end - happened [before](https://github.com/QuickSander/homebridge-mieleathome/issues/91#issuecomment-2558620582) (there are several other issues that appear to be at Miele's end). Usually these are resolved within a day, so you might want to try again.
2. Error 401 is, probably, an "unauthorized" error, so check that the Client ID and the Client Secret do not have any leading or trailing spaces and, if you have a picture, that you have entered them correctly.
3. I think that the first time I tried to set all this up, I missed to check the "I agree to terms of use." check-box in step 3 and that caused some issues later (but I don't recall what exactly happened).
4. I don't recall why, but I think that it was important to do step 6 - restart Homebridge - for the setup to work.

Hope this helps.

**Reply by mark-fitzgerald** [reply]
Thank you for the additional suggestions. I've tried all those items, and still no connection. I'll post in the issues section.
